core/blockchain.go

```diff
@@ -597,14 +603,14 @@ func (self *BlockChain) Rollback(chain []common.Hash) {
 		hash := chain[i]
 
 		if self.hc.CurrentHeader().Hash() == hash {
-			self.hc.SetCurrentHeader(self.GetHeader(self.hc.CurrentHeader().ParentHash))
+			self.hc.SetCurrentHeader(self.GetHeader(self.hc.CurrentHeader().ParentHash, self.hc.CurrentHeader().Number.Uint64()-1))
```

Comment:
Although we can query the current header here multiple times consecutively safely because the blockchain's mutex is locked, it looks and feels wrong and a code refactor could break it. Please fetch the current header once and access it fields for both the hash and number.